### DISCUSSION TITLE
Update form state inside onChange based on which checkbox the user has selected

### DISCUSSION THREAD

**Discussion by RSchneider94** [react-hook-form/react-hook-form#9151] [Q&A] [answered]
Hey folks,

[Sandbox](https://codesandbox.io/s/peaceful-rgb-zvhq3v?file=/src/App.tsx)

I have to take an action based on which checkbox the user has chosen and then update the form based on that. So I have a form that looks like this:

![image](https://user-images.githubusercontent.com/11929659/193948735-490e90ae-beb6-4f86-9711-fccf9df97e8d.png)
 
Then, whenever the user selects, e.g. "Manage" in case it has "View" selected before, I need to deselect "View" because with "Manage" user already has "View" right. I think you know what I mean. I have tried to make use of `useEffect` + `watch` + `setValue` but as the react-hooks linter warns about the need to add them as deps, you know what happens, infinite loop. So I'm wondering how it would be possible to achieve what I need. Because as the docs states, for the `onChange` inside `Controller`:

> This prop update [formState](https://react-hook-form.com/api/useform/formstate) and you should avoid manually invoke [setValue](https://react-hook-form.com/api/useform/setvalue) or other API related to field update

Any help would be very welcome!

Thanks!

**Comment by leapful** [comment, marked as answer]
@RSchneider94 : The logic can be simplified by just using custom "onChange" for CheckboxGroup component.

For example:

```
<CheckboxGroup
                onChange={(e) => {
                  if (
                    e.includes("view-documents") &&
                    e.includes("manage-documents")
                  ) {
                    onChange(["manage-documents"]);
                  } else {
                    onChange(e);
                  }
                }}
                {...newField}
              >
```

Working demo: https://codesandbox.io/s/nervous-saha-80b1ho?file=/src/App.tsx

**Reply by RSchneider94** [reply]
> @RSchneider94 : The logic can be simplified by just using custom "onChange" for CheckboxGroup component.
> 
> For example:
> 
> ```
> <CheckboxGroup
>                 onChange={(e) => {
>                   if (
>                     e.includes("view-documents") &&
>                     e.includes("manage-documents")
>                   ) {
>                     onChange(["manage-documents"]);
>                   } else {
>                     onChange(e);
>                   }
>                 }}
>                 {...newField}
>               >
> ```
> 
> Working demo: https://codesandbox.io/s/nervous-saha-80b1ho?file=/src/App.tsx

hey @leapful , thanks! that's awesome! didn't know that I could send an array in the `onChange` event like that, cool!

**Comment by liel-almog** [comment]
Just to be sure, when creating an input component and using a custom 'onChnage; we need to use the `field.onChange` and pass than our value?

```typescript
const handleChange: AntInputNumberProps['onChange'] = (val: any) => {
        if (isNaN(Number(val))) {
            field.onChange(null);
            return;
        }
        field.onChange(+val);
    };
```

Like this?
